### DISCUSSION TITLE
failing to run my UI after using a new font

### DISCUSSION THREAD

**Discussion by famito** [ocornut/imgui#7043] [New Users Build/Link/Run issues ONLY!]
i have recently bought a C++ UI on internet, i am kinda very new to développement and i dont have a large knowledge of C++.
after a few hours customizing the UI and changing all colors and text etc i now want to add a new font to the UI because the one used is kinda simple. 
with the help of chat gpt iv'e learn a lot and see that i need to use the fonction AddFontFromFileTTF wich is what iv'e done.
the problem is that the old font is loaded from AddFontFromMemoryTTF, after trying a lot of  option to run the new font, none of them have been successful. 
on a copy project iv'e tried to delete the Inter.h include wich is the old font used and all the place where Inter or &inter is used after seing that removing this: 
io.Fonts->AddFontFromMemoryTTF(&inter, sizeof inter, 16 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
tab_text1 = io.Fonts->AddFontFromMemoryTTF(&inter, sizeof inter, 12 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
tab_text2 = io.Fonts->AddFontFromMemoryTTF(&inter, sizeof inter, 24 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
tab_text3 = io.Fonts->AddFontFromMemoryTTF(&inter, sizeof inter, 40 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico = io.Fonts->AddFontFromMemoryTTF(&icon, sizeof icon, 25 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_2 = io.Fonts->AddFontFromMemoryTTF(&Menuicon, sizeof Menuicon, 20 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_subtab = io.Fonts->AddFontFromMemoryTTF(&icon, sizeof icon, 35 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_logo = io.Fonts->AddFontFromMemoryTTF(&icon, sizeof icon, 31 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
tab_text = io.Fonts->AddFontFromMemoryTTF(&inter, sizeof inter, 19 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_minimize = io.Fonts->AddFontFromMemoryTTF(&icon, sizeof icon, 27 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ImGui::StyleColorsDark();
get me back to default font. 
The problem i encounter now is that when i click on 'run without debug" the UI examples dont appear and let me with a white screen that close after a second. 
my font seems to be correctly loaded with: 

void InitializeImGui() {
    // Charger la nouvelle police depuis un fichier TTF
    const char* newFontPath = "C:\\Users\\Famito\\Desktop\\police\\Pacifico.ttf";
    ImGuiIO& io = ImGui::GetIO();
    ImFont* newFont = io.Fonts->AddFontFromFileTTF(newFontPath, 1);
    if (newFont) {
    // La nouvelle police a été chargée avec succès
    printf("Nouvelle police chargée avec succès!\n");
}
else {
    // Échec du chargement de la nouvelle police
    printf("Échec du chargement de la nouvelle police.\n");
}
ImGui::GetIO().FontDefault = newFont;
and after that iv'e used : ImGui::PushFont(newFont); before and ImGui::PopFont(); after all fonction where ImGui and a text is added. 
lmk if any other information/screeshot or spécification is needed. 
THANKS FOR THOSE WHO READ THIS ! 
![debut](https://github.com/ocornut/imgui/assets/151673740/a877a9a7-4f75-4b59-99d3-39a0a9cea84f)
![Capture](https://github.com/ocornut/imgui/assets/151673740/e056110d-9471-40cd-aeab-5ae45a1c8844)
![dqd](https://github.com/ocornut/imgui/assets/151673740/7204fb82-a616-4ff0-94be-935338fd5526)

**Comment by sakiodre** [comment]
There are quite a few problems with your code, I suggest looking at the example code to learn the structure of Dear ImGui.

### Problem 1
> when i click on 'run without debug" the UI examples dont appear and let me with a white screen that close after a second

That means it crashed, don't "run without debug", simply hit F5 and attach the debugger. If you did that, you could see that it'll throw an error in imgui_draw.cpp:
```cpp
IM_ASSERT(!Locked && "Cannot modify a locked ImFontAtlas between NewFrame() and EndFrame/Render()!");
```
The error is pretty self-explanatory, in your second picture, you tried to add a font between `NewFrame` and `EndFrame`, this is illegal. You don't need to add the font again because you already added it in the function `IntializeImGui`. Also, you loaded your font with a size of `1`, I'm not sure how that will work out, but just make it a reasonable size, like `16`.


### Problem 2
The original code uses a lot of fonts, which are `tab_text1`, `tab_text2`, `ico_subtab`, etc... You can't just remove them and expect it to work.
If those fonts weren't initialized, *(i.e. not set a value)*, it'd be an undefined behavior. However, if those are `nullptr` *(i.e. ZERO)*, ImGui will just use the default font.

You must initialize them all, or remove what uses them. I suggest adding this, just to make it runnable:
```cpp
ImFont* newFont = io.Fonts->AddFontFromFileTTF("C:\\Users\\Famito\\Desktop\\police\\Pacifico.ttf", 16);
tab_text1 = nullptr;
tab_text2 = nullptr;
tab_text3 = nullptr;
ico = nullptr;
ico_2 = nullptr;
ico_subtab = nullptr;
ico_logo = nullptr;
tab_text = nullptr;
ico_minimize = nullptr;
```
Once that works, you should initialize them with the fonts you want, and optionally remove fonts you don't intend to use and their related code.

### Problem 3
You correctly used `PushFont` and `PopFont`, but in the third image, you can see that the font `tab_text3` is being passed to the function `AddText`, which will override the `PushFont`.

**Reply by famito** [reply]
okay thanks for all that info, iv'e by mistake try to use my old copy by just replacing main.cpp and keeping all the include and this basically fuck everything.
So, iv'e take back my last working copy and trying to start back correctly. 
I should just add: 
```
ImFont* newFont = io.Fonts->AddFontFromFileTTF("C:\\Users\\Famito\\Desktop\\police\\Pacifico.ttf", 16);
tab_text1 = nullptr;
tab_text2 = nullptr;
tab_text3 = nullptr;
ico = nullptr;
ico_2 = nullptr;
ico_subtab = nullptr;
ico_logo = nullptr;
tab_text = nullptr;
ico_minimize = nullptr; 
```
and then try to use :  ImGui::PushFont(newFont); before and ImGui::PopFont(); after for all the part where a text is added ?
Thanks again for ur help

**Reply by famito** [reply]
"update"
after copying what u send i now understand how it work and used it like this and it work
```
tab_text1 = nullptr;
tab_text2 = nullptr;
tab_text3 = nullptr;
const char* fontFilePath = "C:\\Users\\Famito\\Desktop\\police\\stocky.ttf";
ImFont* tab_text1 = io.Fonts->AddFontFromFileTTF(fontFilePath, 19 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ImFont* tab_text2 = io.Fonts->AddFontFromFileTTF(fontFilePath, 19 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ImFont* tab_text3 = io.Fonts->AddFontFromFileTTF(fontFilePath, 19 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico = io.Fonts->AddFontFromMemoryTTF(&icon, sizeof icon, 25 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_2 = io.Fonts->AddFontFromMemoryTTF(&Menuicon, sizeof Menuicon, 20 * dpi_scale, NULL, io.Fonts->GetGlyphRangesCyrillic());
ico_subtab = nullptr;
ico_logo = nullptr;
tab_text = nullptr;
ico_minimize = nullptr;
```
THANKS FOR UR HELP <3

**Reply by sakiodre** [reply]
By doing `ImFont* tab_text1 = ...`, you just create a local variable instead of setting the value for the global one. You can see this by the color of it, global variables are light gray, and local variables are light blue. The correct code would be `tab_text1 = ...`. In most IDE, you can right-click a variable and "Go to definition" to see where it was declared.

One more tip, it is considered a better practice to use `const char fontFilePath[]` instead of `const char* fontFilePath`

**Reply by famito** [reply]
Okay, thanks I will try it tomorrow.
And that’s a choice to not using it globally, I want to add a font for each tab_text.

My next goal is to change the texture ID of my ESP preview, already got the jpeg without background i just need to see how to convert it into TextureID. 
lmk if u know ab this and if u got any tips or suggestions.
The actual one is like this (all font and Imvec2 are not setup it’s just test 😅)
![IMG_2953](https://github.com/ocornut/imgui/assets/151673740/46cab3ec-b53f-492c-ad50-439a1eda5130)
